crowdsourcehinter/static/js/src/crowdsourcehinter.js

```diff
@@ -0,0 +1,305 @@
+                $(function(){
+                    var data = {
+                        hint: hint
+                    };
+                    html = Mustache.render($("#show_hint_feedback").html(), data);
+                });
+                $(this).append(html);
+                var html = "";
+                var template = $('#add_hint_creation').html();
+                var data = {};
+                html = Mustache.render(template, data);
+                $(this).append(html);
+            }
+        });
+    }
+
+    /**
+     * Show options to remove or return reported hints from/to the hint pool. Called after
+     * correctly answering the question, only visible to staff.
+     * @param result is the reported hint text
+     */
+    function showReportedFeedback(result){
+        var html = "";
+        $(function(){
+            var template = $('#show_reported_feedback').html();
+            var data = {
+                hint: result
+            };
+            html = Mustache.render(template, data);
+        });
+        $(".csh_reported_hints", element).append(html);
+    }
+
+    /**
+     * Append new divisions into html for each answer the student submitted before correctly 
+     * answering the question. showHintFeedback appends new hints into these divs.
+     * When the hinter is set to show best, only one div will be created.
+     * @param student_answers is the text of the student's incorrect answer
+     */
+    function showStudentSubmissionHistory(student_answers){
+        var html = "";
+        var template = $('#show_answer_feedback').html();
+        var data = {
+            answer: student_answers
+        };
+        html = Mustache.render(template, data);
+        $(".csh_feedback", element).append(html);
+    }
+
+    /**
+     * Set up student feedback on hints and contribution of new hints. Incorrect answer(s) and the
+     * the corresponding hint(s) shown to the student are displayed. Students can upvote/downvote/report
+     * hints or contribute a new hint for their incorrect answer(s).
+     * Only one incorrect answer and hint will be shown when the hinter is set to show best.
+     * @param result is a dictionary of incorrect answers and hints, with the index being the hint and the value
+     * being the incorrect answer
+     */
+    function showStudentContribution(result){
+    //Set up the student feedback stage. Each student answer and all answer-specific hints for that answer are shown
+    //to the student, as well as an option to create a new hint for an answer.
+        if(data.isStaff){
+            $('.crowdsourcehinter_block', element).attr('class', 'crowdsourcehinter_block_is_staff');
+            $.each(result, function(index, value) {
+                if(value == "Reported") {
+                    //index represents the reported hint's text
+                    showReportedFeedback(index);
+                }
+            });
+        }
+        if(!isShowingHintFeedback){
```

Comment:
Why do we need this? 

1. Does this ever get called twice? 
2. If this was very well structured, calling it twice would simply not change anything.